src/cache/cache_las.c

```diff
@@ -536,26 +642,232 @@ __wt_las_remove_block(WT_SESSION_IMPL *session,
+	 * sweep.
+	 */
+	if (key->size != 0) {
+		__wt_cursor_set_raw_key(cursor, key);
+		ret = cursor->search_near(cursor, &notused);
+
+		/*
+		 * Don't search for the same key twice; if we don't set a new
+		 * key below, it's because we've reached the end of the table
+		 * and we want the next pass to start at the beginning of the
+		 * table. Searching for the same key could leave us stuck at
+		 * the end of the table, repeatedly checking the same rows.
+		 */
+		key->size = 0;
+	} else
+		ret = __las_sweep_init(session);
+
+	if (ret != 0)
+		goto srch_notfound;
+
+	/*
+	 * The sweep server wakes up every 10 seconds (by default), it's a slow
+	 * moving thread. Try to review the entire lookaside table once every 5
+	 * minutes, or every 30 calls.
+	 *
+	 * The reason is because the lookaside table exists because we're seeing
+	 * cache/eviction pressure (it allows us to trade performance and disk
+	 * space for cache space), and it's likely lookaside blocks are being
+	 * evicted, and reading them back in doesn't help things. A trickier,
+	 * but possibly better, alternative might be to review all lookaside
+	 * blocks in the cache in order to get rid of them, and slowly review
+	 * lookaside blocks that have already been evicted.
+	 */
+	cnt = (uint64_t)WT_MAX(100, cache->las_entry_count / 30);
+
+	/* Walk the file. */
+	for (; cnt > 0 && (ret = cursor->next(cursor)) == 0; --cnt) {
+		/*
+		 * Give up if the cache is stuck: we are ignoring the cache
+		 * size while scanning the lookaside table, so we're making
+		 * things worse.
+		 */
+		if (__wt_cache_stuck(session))
+			cnt = 1;
+
+		/*
+		 * If the loop terminates after completing a work unit, we will
+		 * continue the table sweep next time. Get a local copy of the
+		 * sweep key, we're going to reset the cursor; do so before
+		 * calling cursor.remove, cursor.remove can discard our hazard
+		 * pointer and the page could be evicted from underneath us.
+		 */
+		if (cnt == 1) {
+			WT_ERR(__wt_cursor_get_raw_key(cursor, key));
+			if (!WT_DATA_IN_ITEM(key))
+				WT_ERR(__wt_buf_set(
+				    session, key, key->data, key->size));
+		}
+
+		WT_ERR(cursor->get_key(cursor,
+		    &las_pageid, &las_id, &las_counter, &las_key));
+
+		/*
+		 * If the entry belongs to a dropped tree, discard it.
+		 *
+		 * Cursor opened overwrite=true: won't return WT_NOTFOUND
+		 * should another thread remove the record before we do (not
+		 * expected for dropped trees), and the cursor remains
+		 * positioned in that case.
+		 *
+		 * TODO it would also be good to remove entries in lookaside
+		 * from live files that have aged out.  If we track for each
+		 * entry whether it was the on-page value chosen by
+		 * reconciliation, we can safely remove entries from that point
+		 * on (for the given key) that are visible to all readers.
+		 */
+		if (__bit_test(cache->las_sweep_dropmap,
+		    las_id - cache->las_sweep_dropmin)) {
+			WT_ERR(cursor->remove(cursor));
+			++remove_cnt;
+		}
```

Comment:
Note the TODO above: I'd like a followup ticket to follow close after this.

Reply:
Done: WT-3737